### ISSUE TITLE
INSERT ... SET ... RETURNING returns null

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3140]
It appears that this doesn't work:

``` java
TodosRecord persisted = jooq.insertInto(TODOS)
                .set(TODOS.CREATION_TIME, currentTime)
                .set(TODOS.DESCRIPTION, todo.getDescription())
                .set(TODOS.ID, id)
                .set(TODOS.MODIFICATION_TIME, currentTime)
                .set(TODOS.TITLE, todo.getTitle())
                .returning()
                .fetchOne();

        //persisted is always null
```

... whereas this does:

``` java
TodosRecord persisted = jooq.insertInto(TODOS)
                .set(createRecord(todo))
                .returning()
                .fetchOne();

        //Persisted is found
```

---

See also:
https://groups.google.com/forum/#!topic/jooq-user/fOsaaZq44Ac

**Comment by viki85in** [comment]
The issue is still reproducible though. I am using mysql with jooq.

**Comment by lukaseder** [comment]
@viki85in Thank you very much for reporting. Can you provide me with the DDL to create the `SUBSCRIBER` table, as well as your MySQL version and MySQL JDBC driver version?

**Comment by viki85in** [comment]
@lukaseder I figured out the problem. And I think a lot of developers have bumped into the same scenario. I had added the identify auto increment after generating the objects. 

Once I regenerated the objects, the issue is not reproducible. Sorry for the confusion. My bad.

**Comment by lukaseder** [comment]
Oh I see. Yes, jOOQ needs the "identity" information on the right column, otherwise it won't be able to fetch the right identity value (in MySQL)

Thanks a lot for the feedback!

**Comment by tmjeee** [comment]
Any chance this going into 3.9.x branch?

**Comment by lukaseder** [comment]
Hi @tmjeee: Thanks for your message. I'm not sure what you mean - this issue was resolved as "worksforme" - there is currently no fix. Maybe, you experienced a similar issue? Would you mind creating a new issue on GitHub?

**Comment by txsmith** [comment]
As discussed in #2123, I'm trying to get `INSERT .. RETURNING` working with `ON DUPLICATE KEY`.
@lukaseder Suggested to cast the type returned by `InsertOnDuplicateStep.onDuplicateKeyIgnore()` to `InsertReturningStep<R>` and call `returning()` on that, but when I do that the query always returns a `null` record.

To demonstrate I created a [repo](https://github.com/txsmith/jooq-minimal-example) with one test that describes my problem.

I'd love your advice on this!

**Comment by deinspanjer** [comment]
I am stuck really needing this capability as well, and seeing the same result.  I turned on debug logging, and the resulting INSERT statement ends with `on conflict do nothing`, it does not include any `returning` clause.

**Comment by lukaseder** [comment]
@txsmith: I'm not sure if this is really the same problem, although, if it is, then it's not a problem. Your test case is simply incorrect:

```java
// Insert the first entry (if not exists)
db.insertInto(PERSON, PERSON.FIRST_NAME, PERSON.LAST_NAME)
        .values("Thomas", "Smith")
        .onDuplicateKeyIgnore()
        .execute();

Integer records = db.selectCount().from(PERSON).fetchOneInto(Integer.class);

assertThat(records).isOne();

// Insert the duplicate entry
InsertReturningStep<?> query = (InsertReturningStep<?>)
        db.insertInto(PERSON, PERSON.FIRST_NAME, PERSON.LAST_NAME)
        .values("Thomas", "Smith")
        .onDuplicateKeyIgnore();

Record result = query.returning(PERSON.FIRST_NAME).fetchOne();
assertThat(result).isNotNull();
```

If you run the second query in PostgreSQL directly, you'll get an empty result set:

```sql
insert into "example"."person" (
  "first_name", 
  "last_name"
)
values (
  'Thomas', 
  'Smith'
)
on conflict do nothing
returning "example"."person"."first_name"
```

Yielding...

![image](https://user-images.githubusercontent.com/734593/27304564-b53b1cda-553e-11e7-959a-1d81707e26f0.png)

`fetchOne()` in jOOQ consistently returns `null` if there was no record, throughout the jOOQ API. Perhaps, `fetchOptional()` is more aligned with your expectations?

@deinspanjer: I'm happy to look into any steps to reproduce this. The test case @txsmith provided does generate that returning clause...

**Comment by txsmith** [comment]
Right! I should have seen that.. Thanks for your help!

**Comment by txsmith** [comment]
Right, so it turns out that I got the query wrong in the first place. My intention was to have the query *always* return the `first_name` regardless of any conflicts. As you pointed out, this is not the way `on conflict do nothing returning ...` works, so I switched to `on conflict do update ... returning ...`:

```java
// Insert the duplicate entry
InsertReturningStep<?> query = (InsertReturningStep<?>)
        db.insertInto(PERSON, PERSON.FIRST_NAME, PERSON.LAST_NAME)
        .values("Thomas", "Smith")
        .onConflict(PERSON.FIRST_NAME, PERSON.LAST_NAME)
        .doUpdate()
        .set(PERSON.FIRST_NAME, "Thomas");

Record result = query.returning(PERSON.FIRST_NAME).fetchOne();
```
It seems like kind of a hack on the sql side, but at least this works! Thanks for pushing me in the right direction :)

**Comment by deinspanjer** [comment]
I was basing my code off of txsmith's work, so let me try again and if I can reproduce a failing case I'll share it.

I was wanting to do the same thing as him, return the id of the existing item if I try to insert a duplicate, otherwise, insert and return the new id.

**Comment by jakesyl** [comment]
Thanks so much @txsmith, this worked for me. Would still love to see a PR adding .returning() to `InsertOnDuplicateSetStep` though.
